modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -0,0 +1,117 @@
+import org.apache.ignite.internal.table.distributed.raft.snapshot.SnapshotUri;
+import org.apache.ignite.internal.table.message.SnapshotMetaRequest;
+import org.apache.ignite.internal.table.message.SnapshotMetaResponse;
+import org.apache.ignite.internal.table.message.TableMessagesFactory;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.raft.jraft.entity.RaftOutter.SnapshotMeta;
+import org.apache.ignite.raft.jraft.storage.SnapshotStorage;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotCopier;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+
+/**
+ * Snapshot copier implementation for partitions. Used to stream partition data from the leader to the local node.
+ */
+public class IncomingSnapshotCopier extends SnapshotCopier {
+    /** Messages factory. */
+    private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
+
+    /** {@link SnapshotStorage} instance for the partition. */
+    private final PartitionSnapshotStorage snapshotStorage;
+
+    /** Snapshot URI. */
+    private final SnapshotUri snapshotUri;
+
+    /** Rebalance thread-pool, used to write data into a storage. */
+    //TODO Use external pool.
+    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+
+    /**
+     * Snapshot meta read from the leader.
+     *
+     * @see SnapshotMetaRequest
+     */
+    private SnapshotMeta snapshotMeta;
+
+    /**
+     * Constructor.
+     *
+     * @param snapshotStorage Snapshot storage.
+     * @param snapshotUri Snapshot URI.
+     */
+    public IncomingSnapshotCopier(PartitionSnapshotStorage snapshotStorage, SnapshotUri snapshotUri) {
+        this.snapshotStorage = snapshotStorage;
+        this.snapshotUri = snapshotUri;
+    }
+
+    @Override
+    public void cancel() {
+        //TODO Implement.
+    }
+
+    @Override
+    public void join() throws InterruptedException {
+        //TODO Implement proper join.
+        threadPool.awaitTermination(1, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void start() {
+        ClusterNode sourceNode = snapshotStorage.topologyService.getByConsistentId(snapshotUri.nodeName);
+
+        MessagingService messagingService = snapshotStorage.outgoingSnapshotsManager.messagingService();
+
+        threadPool.submit(() -> {
+            //TODO https://issues.apache.org/jira/browse/IGNITE-17262
+            // Following code is just an example of what I expect and shouldn't be considered a template.
+            CompletableFuture<NetworkMessage> metaRequestFuture = messagingService.invoke(
+                    sourceNode,
+                    MSG_FACTORY.snapshotMetaRequest().id(snapshotUri.snapshotId).build(),
+                    1000L
+            );
+
+            metaRequestFuture.whenCompleteAsync((networkMessage, throwable) -> {
```

Comment:
Why `whenCompleteAsync`? You simply do a field assignment here

Reply:
This is not a final code, but I'll change it if you wish